projects/swimlane/ngx-ui/src/lib/components/calendar/calendar.component.ts

```diff
@@ -122,31 +210,138 @@ export class CalendarComponent implements OnInit, ControlValueAccessor {
+
   ngOnInit(): void {
     this.activeDate = this.createMoment(this.value);
     this.weeks = getMonth(this.activeDate);
+    this.monthsList = moment.monthsShort();
+    this.current = this.activeDate;
+    this.startYear = this.current.year() - (this.current.year() % 20) + 1;
+    this.validateView();
   }
 
   getDayActive(date: moment.Moment): boolean {
     return date.isSame(this.value, 'day');
   }
 
-  getDayDisabled(date: moment.Moment): boolean {
-    if (this.disabled) return true;
-    if (!date) return false;
+  getMonthActive(month: string): boolean {
+    const date = moment(this.value).clone().month(month);
+    return date.isSame(this.value, 'month') && date.isSame(this.activeDate, 'year');
+  };
 
-    const isBeforeMin = this.minDate && date.isBefore(this.parseDate(this.minDate));
-    const isAfterMax = this.maxDate && date.isAfter(this.parseDate(this.maxDate));
+  getCurrentMonth(month: string): boolean {
+    const date = this.activeDate.clone().month(month);
+    return date.isSame(this.current, 'month') && date.isSame(this.current, 'year') && !date.isSame(this.value, 'month');
+  }
+
+  getYearActive(year: number): boolean {
+    const date = moment(this.value).clone().year(year);
+    return date.isSame(this.value, 'year');
+  }
+
+  getCurrentYear(year: number): boolean {
+    const date = this.activeDate.clone().year(year);
+    return date.isSame(this.current, 'year') && !date.isSame(this.value, 'year');
+  }
+
+  getDisabled(value: any, type: string): boolean {
+    if (this.disabled) return true;
+    if (!value) return false;
 
+    let date;
+    switch(type) {
+      case 'day':
+        date = value;
+        break;
+      case 'month':
+        date = this.activeDate.clone().month(value);
+        break;
+      case 'year':
+        date = this.activeDate.clone().year(value);
+        break;
+      default:
+        return false;
+    }
+    const isBeforeMin = this.minDate && date.isBefore(this.parseDate(this.minDate), type);
+    const isAfterMax = this.maxDate && date.isAfter(this.parseDate(this.maxDate), type);
     return isBeforeMin || isAfterMax;
   }
 
   onDayClick(day: CalenderDay): void {
-    this.value = day.date.clone().toDate();
-    if (day.prevMonth) {
-      this.prevMonth();
-    } else if (day.nextMonth) {
-      this.nextMonth();
+    this.activeDate = day.date.clone();
+    this.value = this.activeDate.toDate();
+    if (day.prevMonth || day.nextMonth) {
+      this.weeks = getMonth(this.activeDate);
+    }
+  }
+
+  onMonthClick(month: string): void {
+    this.activeDate.month(month);
+    this.value = this.activeDate.toDate();
+    if (this.minView !== 'month') {
+      this.currentView = 'date';
+      this.weeks = getMonth(this.activeDate);
+    }
+  }
+
+  onYearClick(year: number): void {
+    this.activeDate.year(year);
+    this.value = this.activeDate.toDate();
+    if (this.minView !== 'year') {
+      this.currentView = 'month';
+      this.weeks = getMonth(this.activeDate);
```

Comment:
Is this needed?

Reply:
That is for switching views down. (e.g. years view selected -> months view show)